Cleanup some logic in pickler

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22953 Add initial support for serializing classes
* #22883 Remove unneeded IValue copy in unpickler.
* **#22882 Cleanup some logic in pickler**

As part of adding object serialization, this PR cleans up logic in
the pickler encoding:

* Memoization is centralized in one function. This makes it harder to
  accidentally forget to add pushMemoization
* Strings that are not IValues are memoized. This should prevent things
  like the string "storage" from being repeated for each Tensor.
* Some renaming to make the difference between strings, globals,
  and string encoding functions more clear.

Differential Revision: [D16270332](https://our.internmc.facebook.com/intern/diff/D16270332)

zdevito committed:
Cleanup some logic in pickler

As part of adding object serialization, this PR cleans up logic in
the pickler encoding:

* Memoization is centralized in one function. This makes it harder to
  accidentally forget to add pushMemoization
* Strings that are not IValues are memoized. This should prevent things
  like the string "storage" from being repeated for each Tensor.
* Some renaming to make the difference between strings, globals,
  and string encoding functions more clear.

driazati commented on torch/csrc/jit/pickler.h:
```diff
@@ -183,9 +182,17 @@ class Pickler {
+  // map we need to keep all the memoized values alive during the pickle.
+  // Otherwise, it is possible that a raw address gets reused for another
+  // object, and we will alias it to the old object at that address.
+  std::vector<IValue> memoized_ivalues_;
```
This seems redundant with `memoized_ivalue_map_`, it could be 1 map of `std::unordered_map<IValue, uint32_t>` with a custom hash that calls `internalToPointer()`

zdevito replied:
This is just a rename of what was there before. I feel like figuring out what the custom hash is doing is harder to understand than this.

driazati commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -256,28 +256,36 @@ void Pickler::pushBinGet(uint32_t memo_id) {
+  }
+}
+
+void Pickler::pushRawStringData(const std::string& string) {
```
Maybe a better name is `pushBytes()` since that's all it's actually doing

zdevito committed:
Update on "Cleanup some logic in pickler"

Cleanup some logic in pickler

As part of adding object serialization, this PR cleans up logic in
the pickler encoding:

* Memoization is centralized in one function. This makes it harder to
  accidentally forget to add pushMemoization
* Strings that are not IValues are memoized. This should prevent things
  like the string "storage" from being repeated for each Tensor.
* Some renaming to make the difference between strings, globals,
  and string encoding functions more clear.

gh-metadata: pytorch pytorch 22882 gh/zdevito/71/head
